sdk/core/azure-core/src/http/curl/curl_session_private.hpp

```diff
@@ -418,6 +424,13 @@ namespace Azure { namespace Core { namespace Http {
      * @return The size of the payload.
      */
     int64_t Length() const override { return m_contentLength; }
+
+    /**
+     * @brief Return the network connection if the server indicated that the connection is upgraded.
+     *
+     * @return The network connection, or null if the connection was not upgraded.
+     */
+    std::unique_ptr<CurlNetworkConnection> GetUpgradedConnection();
```

Comment:
we use the prefix `Extract` instead of `Get` to indicate the change of ownership.  So, this should be `ExtractUpgradedConnection()`.

You could even remove the concept of `upgrading` and just extract the connection from the session when you know the transport supports websockets.  Then you let the session go out of scope and check if the connection is gone, then do not call move to pool. This way the session doesn't really need to know about anything about the protocol (upgrading is a term associated with the protocol and not really to a session).

In terms of the session, it only needs to support Extracting the connection at any time, and the caller (the protocol implementation would know when to extract the connection)